GVRf/Extensions/gvrf-particlesystem/src/main/java/org/gearvrf/particlesystem/GVREmitter.java

```diff
@@ -0,0 +1,365 @@
+ * objects each have a mesh attached to them, with the number of vertices in the mesh
+ * being equal to the emit rate of the emitter, i.e., every second a new scene object with a mesh
+ * having an emitRate number of vertices is added to the emitter as a child.
+ * These vertices act as the particles of the system. Consequently, all those
+ * children objects which have exceeded their age limits are deleted every second.
+ *
+ */
+
+class GVREmitter extends GVRSceneObject {
+
+    private int MAX_EMIT_RATE = 500;
+
+    protected int mEmitRate = 300;
+    protected boolean mEnableEmitter = true;
+    protected GVRContext mGVRContext = null;
+
+
+    //list of active scene objects containing particle meshes,
+    // and their corresponding spawn times.
+    protected ArrayList<Pair<GVRSceneObject, Float>> meshInfo = null;
+
+
+    //particle properties
+    protected float mMaxAge = 1.5f;
+    protected float mParticleSize = 50.0f;
+    protected float minVelocity = 1.5f;
+    protected float maxVelocity = 4.5f;
+    protected Vector3f mEnvironmentAcceleration ;
+    private Vector4f mColor;
+    private float mParticleSizeRate = 0.0f;
+    private boolean mFadeWithAge = false;
+    private GVRTexture mParticleTexture;
+
+    //attributes of the particles generated every second.
+    protected float[] mParticlePositions;
+    protected float[] mParticleVelocities;
+    protected float [] mParticleGenTimes;
+
+    protected boolean burstMode = false;
+    private boolean executeOnce = true;
+
+    //values corresponding to the particles which degine the bounding volume of the system.
+    private float[] particleBoundingVolume;
+    private float[] BVSpawnTimes;
+    private float[] BVVelocities;
+
+    private float currTime = 0;
+
+    public GVREmitter(GVRContext gvrContext)
+    {
+        super(gvrContext);
+        mGVRContext = gvrContext;
+        meshInfo = new ArrayList<Pair<GVRSceneObject, Float>>();
+        mEnvironmentAcceleration = new Vector3f(0.0f,0.0f,0.0f);
+        mColor = new Vector4f(1.0f, 1.0f, 1.0f, 1.0f);
+        setParticleVolume(100,100,100);
+    }
+
+    /**
+     * Iterate over the active meshes and delete the ones whose every particle
+     * has exceeded the maximum age limit.
+     * <p>
+     * Particle meshes are generated for upto one second in the future.
+     * Therefore the maximum time particle mesh should be allowed to remain
+     * in the scene is (maxAge of the particles + one second).
+     * <p>
+     */
+
+    protected void onDrawFrame() {
+        int numActiveMeshes = meshInfo.size();
+        for (int i = 0; i < numActiveMeshes; i++)
+        {
+            float currMeshSpawnTime = meshInfo.get(i).second;
+            if (currTime - currMeshSpawnTime > mMaxAge + 1) {
+                GVRSceneObject toDelete = meshInfo.get(i).first;
+                GVREmitter.this.removeChildObject(toDelete);
+                meshInfo.remove(i);
```

Comment:
You are removing something from the collection so the size changes. But you still iterate over the old size of the collection. This is the source of the exceptions I mentioned. Seem to be non-critical but ideally you want it fixed.